src/engine/mod.rs

```diff
@@ -0,0 +1,320 @@
+
+                Ok(Number::from_f64(validate_f64(result)?).unwrap())
+            }
+            MathOperator::Division => {
+                if lhs.is_i64() && rhs.is_i64() {
+                    // Try to divide integers and if there's no remained, return result as integer as well
+                    if let Some(0) = lhs.as_i64().unwrap().checked_rem(rhs.as_i64().unwrap()) {
+                        if let Some(x) = lhs.as_i64().unwrap().checked_div(rhs.as_i64().unwrap()) {
+                            return Ok(Number::from(x));
+                        }
+                    }
+                }
+
+                let lhs = lhs.as_f64().unwrap();
+                let rhs = rhs.as_f64().unwrap();
+                let result = lhs / rhs;
+
+                Ok(Number::from_f64(validate_f64(result)?).unwrap())
+            }
+        }
+    }
+
+    fn eval_args(&self, args: &HashMap<String, Expression>, context: &Context) -> Result<HashMap<String, Value>> {
+        let mut result = HashMap::new();
+
+        for (k, v) in args.iter() {
+            result.insert(k.to_string(), self.eval(v, context)?);
+        }
+
+        Ok(result)
+    }
+
+    fn eval_function(&self, name: &str, args: &HashMap<String, Expression>, context: &Context) -> Result<Value> {
+        let args = self.eval_args(args, context)?;
+
+        let f = self.functions.get(name);
+
+        if f.is_none() {
+            bail!("function `{}` not found", name);
+        }
+
+        (f.unwrap())(&args)
+    }
+
+    fn eval_filter(&self, name: &str, value: &Value) -> Result<Value> {
+        let f = self.filters.get(name);
+
+        if f.is_none() {
+            bail!("filter `{}` not found", name);
+        }
+
+        (f.unwrap())(value)
+    }
+
+    fn eval_value_as_number(&self, value: &ExpressionValue, context: &Context) -> Result<Number> {
+        let number = match value {
+            ExpressionValue::Integer(x) => Number::from(*x),
+            ExpressionValue::Float(x) => Number::from_f64(*x).unwrap(),
+            ExpressionValue::Identifier(x) => match context.lookup_variable(x)? {
+                Value::Number(ref x) => x.clone(),
+                _ => bail!("identifier does not evaluate to a number"),
+            },
+            ExpressionValue::Math(MathExpression {
+                ref lhs,
+                ref rhs,
+                ref operator,
+            }) => {
+                let lhs = self.eval_as_number(lhs, context)?;
+                let rhs = self.eval_as_number(rhs, context)?;
+                self.eval_math(&lhs, &rhs, *operator)?
+            }
+            ExpressionValue::FunctionCall(FunctionCall { ref name, ref args }) => {
+                match self.eval_function(name, args, context)? {
+                    Value::Number(x) => x,
+                    _ => bail!("result of `{}` is not a number", name),
+                }
+            }
+            ExpressionValue::Boolean(_) => bail!("unable to evaluate boolean as a number"),
```

Comment:
Curious - would it be possible to express this in the type system ? E.g. make the function only take things that it can work on, like introducing a 'smaller' type ? Or would that spin into to much work in other places ?

Reply:
It will, but the AST will be much more complicated and there will be lot of other work to do. Which isn't worth IMHO. Especially now when I still do expect grammar changes.